File: Saver/VaporwaveRenderer.swift
import ScreenSaver
import AppKit

extension AlbumCoverSaverView {

    /// Neon horizon, chrome type, endless grid.
    func drawVaporwave() {
        let W = bounds.width, H = bounds.height
        let pal = featPalette
        let horizon = H * 0.46

        let magenta = NSColor(red: 0.98, green: 0.20, blue: 0.62, alpha: 1)
        let cyan = NSColor(red: 0.25, green: 0.94, blue: 0.98, alpha: 1)
        let violet = NSColor(red: 0.24, green: 0.08, blue: 0.36, alpha: 1)

        // Sky.
        NSGradient(colors: [NSColor(red: 0.05, green: 0.02, blue: 0.14, alpha: 1),
                            violet, magenta.withAlphaComponent(0.75)])?
            .draw(in: NSBezierPath(rect: NSRect(x: 0, y: horizon,
                                                width: W, height: H - horizon)), angle: -90)

        // Sun, banded the way every one of these is.
        let sunR = min(W, H) * 0.20
        let sunC = NSPoint(x: W * 0.5, y: horizon + sunR * 0.62)
        NSGraphicsContext.saveGraphicsState()
        NSBezierPath(ovalIn: NSRect(x: sunC.x - sunR, y: sunC.y - sunR,
                                    width: sunR * 2, height: sunR * 2)).addClip()
        NSGradient(colors: [NSColor(red: 1, green: 0.90, blue: 0.35, alpha: 1), magenta])?
            .draw(in: NSBezierPath(rect: NSRect(x: sunC.x - sunR, y: sunC.y - sunR,
                                                width: sunR * 2, height: sunR * 2)), angle: -90)
        // Slots widen toward the bottom.
        var sy = sunC.y - sunR * 0.10
        var band: CGFloat = 2
        while sy > sunC.y - sunR {
            NSColor(red: 0.10, green: 0.03, blue: 0.20, alpha: 0.9).setFill()
            NSBezierPath(rect: NSRect(x: sunC.x - sunR, y: sy, width: sunR * 2, height: band)).fill()
            sy -= band * 2.4
            band *= 1.25
        }
        NSGraphicsContext.restoreGraphicsState()

        // Ground and perspective grid.
        NSGradient(colors: [NSColor(red: 0.06, green: 0.01, blue: 0.12, alpha: 1),
                            NSColor(red: 0.16, green: 0.02, blue: 0.26, alpha: 1)])?
            .draw(in: NSBezierPath(rect: NSRect(x: 0, y: 0, width: W, height: horizon)),
                  angle: -90)

        cyan.withAlphaComponent(0.75).setStroke()
        for i in -14...14 {
            let g = NSBezierPath()
            g.move(to: NSPoint(x: W * 0.5 + CGFloat(i) * W * 0.02, y: horizon))
            g.line(to: NSPoint(x: W * 0.5 + CGFloat(i) * W * 0.30, y: 0))
            g.lineWidth = 1.4
            g.stroke()
        }
        // Horizontal lines accelerate toward the viewer, so the ground moves.
        let z = CGFloat(phase).truncatingRemainder(dividingBy: 1)
        var k: CGFloat = 0
        while k < 22 {
            let t = (k + z) / 22
            let y = horizon * (1 - t * t)
            let a = max(0, 0.75 - t * 0.6)
            cyan.withAlphaComponent(a).setStroke()
            let g = NSBezierPath()
            g.move(to: NSPoint(x: 0, y: y)); g.line(to: NSPoint(x: W, y: y))
            g.lineWidth = 1.4
            g.stroke()
            k += 1
        }
        magenta.withAlphaComponent(0.9).setFill()
        NSBezierPath(rect: NSRect(x: 0, y: horizon - 1.5, width: W, height: 3)).fill()

        guard albums.indices.contains(featIndex) else { return }

        // Cover floating over the horizon with a neon frame.
        let side = min(H * 0.34, W * 0.26)
        let bob = CGFloat(sin(Double(phase) * 0.7)) * H * 0.008
        let rect = NSRect(x: W * 0.5 - side / 2, y: horizon + H * 0.10 + bob,
                          width: side, height: side)
        NSGradient(colors: [cyan.withAlphaComponent(0.5), cyan.withAlphaComponent(0)])?
            .draw(fromCenter: NSPoint(x: rect.midX, y: rect.midY), radius: side * 0.5,
                  toCenter: NSPoint(x: rect.midX, y: rect.midY), radius: side * 1.15,
                  options: [])
        drawFeaturedCover(in: rect, radius: 2, shadowAlpha: 0.7)
        let fr = NSBezierPath(rect: rect)
        fr.lineWidth = 3
        magenta.setStroke(); fr.stroke()
        fr.lineWidth = 1
        cyan.setStroke(); fr.stroke()

        guard settings.showTrackLabel else { return }
        let text = featText
        let colW = W * 0.8
        let x = W * 0.5 - colW / 2
        let font = fittedFont(text.title.uppercased(), width: colW, maxHeight: H * 0.11,
                              base: max(20, H * 0.052), bold: true)
        // Chromatic fringing: the same word in cyan and magenta, offset.
        let off = max(2, H * 0.004)
        drawText(text.title.uppercased(), x: x - off, top: rect.minY - H * 0.03,
                 width: colW, font: font, color: cyan.withAlphaComponent(0.9),
                 kern: 4, centred: true)
        drawText(text.title.uppercased(), x: x + off, top: rect.minY - H * 0.03,
                 width: colW, font: font, color: magenta.withAlphaComponent(0.9),
                 kern: 4, centred: true)
        let h = drawText(text.title.uppercased(), x: x, top: rect.minY - H * 0.03,
                         width: colW, font: font, color: .white, kern: 4, centred: true)
        drawText(text.artist.uppercased(), x: x, top: rect.minY - H * 0.03 - h - H * 0.014,
                 width: colW, font: retroFont(max(12, H * 0.024)),
                 color: pal.accent, kern: 6, centred: true)
    }
}

File: windows/src/AlbumCoverScreenSaver.Saver/VaporwaveRenderer.cs
using AlbumCoverScreenSaver.Shared;
using SkiaSharp;

namespace AlbumCoverScreenSaver.Saver;

/// <summary>
/// Vaporwave Grid. A sun setting over a neon grid, with the cover floating above
/// the horizon.
/// </summary>
/// <remarks>
/// <para>
/// <b>The one style that does not take its colours from the album.</b> Every
/// other style built so far reads its scheme out of the cover; this one is
/// hard-coded magenta, cyan and violet, because the genre simply requires those
/// colours and a record that happens to be brown does not get a brown vaporwave.
/// The album recolours exactly one line of type, the artist, and nothing else.
/// </para>
/// <para>
/// It also has no idle state to speak of. Nothing about the scene changes when
/// the music stops except the words, so it looks the same running unattended at
/// three in the morning as it does while you are listening.
/// </para>
/// </remarks>
internal sealed class VaporwaveRenderer : IStyleRenderer, IDisposable
{
    private static readonly SKColor Magenta = new(250, 51, 158);
    private static readonly SKColor Cyan = new(64, 240, 250);
    private static readonly SKColor Violet = new(61, 20, 92);
    private static readonly SKColor SkyTop = new(13, 5, 36);
    private static readonly SKColor GroundNear = new(41, 5, 66);
    private static readonly SKColor GroundFar = new(15, 3, 31);
    private static readonly SKColor SunCore = new(255, 230, 89);
    private static readonly SKColor SunSlot = new(26, 8, 51, 230);

    private readonly SaverData _data;
    private readonly AlbumPicker _picker;
    private readonly PaletteStore _palettes;
    private readonly bool _isPreview;

    private readonly Featured _featured;
    private readonly SKPaint _fill = new() { IsAntialias = true };
    private readonly SKPaint _line = new()
    {
        IsAntialias = true,
        Style = SKPaintStyle.Stroke,
        StrokeWidth = 1.4f,
    };

    public VaporwaveRenderer(SaverData data, AlbumPicker picker, PaletteStore palettes, bool isPreview)
    {
        _data = data;
        _picker = picker;
        _palettes = palettes;
        _isPreview = isPreview;
        _featured = new Featured(picker, data.Settings.RecencyBias);
    }

    public void BuildLayout(float width, float height)
    {
        if (_data.Albums.Count == 0) return;

        _featured.Reset(0, _picker.Pick(_data.Albums.Count, _data.Settings.RecencyBias));
        Log.Write($"vaporwave layout: {width:0}x{height:0} points");
    }

    public void Advance(double phase, float width, float height) =>
        _featured.Advance(phase, _data.LiveAlbumIndex, _data.Settings.FeatureSeconds, _data.Albums.Count);

    public void Draw(SKCanvas canvas, float width, float height, double phase)
    {
        var albums = _data.Albums;
        if (albums.Count == 0) return;

        var index = Math.Clamp(_featured.Index, 0, albums.Count - 1);
        var previous = Math.Clamp(_featured.PreviousIndex, 0, albums.Count - 1);
        var palette = _palettes.For(albums[index].Id, _data.ImageFor(index));

        var horizon = height * (1f - VaporwaveGrid.HorizonFraction);
        var sunRadius = Math.Min(width, height) * 0.20f;

        DrawSky(canvas, width, horizon);
        DrawSun(canvas, width, horizon, sunRadius);
        DrawGround(canvas, width, height, horizon);
        DrawRays(canvas, width, height, horizon);
        DrawRungs(canvas, width, height, horizon, phase);

        _fill.Shader = null;
        _fill.Color = Magenta.WithAlpha(230);
        canvas.DrawRect(SKRect.Create(0, horizon - 1.5f, width, 3f), _fill);

        var side = Math.Min(height * 0.34f, width * 0.26f);
        var bob = MathF.Sin((float)phase * 0.7f) * height * 0.008f;

        // Floating above the horizon, not standing on it.
        var rect = SKRect.Create(
            (width * 0.5f) - (side / 2f),
            horizon - (height * 0.10f) - bob - side,
            side, side);

        DrawGlow(canvas, rect, side);

        Drawing.DrawFeaturedCover(
            canvas, _data.ImageFor(index), _data.ImageFor(previous), rect,
            _featured.FadeRaw(phase), radius: 2f, shadowAlpha: 0.7f);

        DrawNeonFrame(canvas, rect);

        if (_data.Settings.ShowTrackLabel && !_isPreview)
        {
            DrawText(canvas, width, height, rect, index, palette);
        }
    }

    private void DrawSky(SKCanvas canvas, float width, float horizon)
    {
        using var shader = SKShader.CreateLinearGradient(
            new SKPoint(0, 0),
            new SKPoint(0, horizon),
            new[] { SkyTop, Violet, Magenta.WithAlpha(191) },
            SKShaderTileMode.Clamp);

        _fill.Shader = shader;
        _fill.Color = SKColors.White;
        canvas.DrawRect(SKRect.Create(0, 0, width, horizon), _fill);
        _fill.Shader = null;
    }

    /// <summary>
    /// The sun, cut across by bands that widen as they go down.
    /// </summary>
    /// <remarks>
    /// The widening is the signature of the genre. Bands of a fixed width give
    /// you a striped circle; bands that grow give you a sun dissolving into the
    /// horizon, which is the whole image.
    /// </remarks>
    private void DrawSun(SKCanvas canvas, float width, float horizon, float radius)
    {
        var centre = new SKPoint(width * 0.5f, horizon - (radius * 0.62f));

        canvas.Save();
        canvas.ClipRect(SKRect.Create(
            centre.X - radius, centre.Y - radius, radius * 2f, radius * 2f));
        canvas.ClipPath(CirclePath(centre, radius), SKClipOperation.Intersect, antialias: true);

        using (var shader = SKShader.CreateLinearGradient(
                   new SKPoint(0, centre.Y - radius),
                   new SKPoint(0, centre.Y + radius),
                   new[] { SunCore, Magenta },
                   SKShaderTileMode.Clamp))
        {
            _fill.Shader = shader;
            _fill.Color = SKColors.White;
            canvas.DrawCircle(centre, radius, _fill);
            _fill.Shader = null;
        }

        _fill.Color = SunSlot;
        foreach (var (down, band) in VaporwaveGrid.SunSlots(radius))
        {
            canvas.DrawRect(
                SKRect.Create(centre.X - radius, centre.Y + down, radius * 2f, band), _fill);
        }

        canvas.Restore();
    }

    private static SKPath CirclePath(SKPoint centre, float radius)
    {
        var path = new SKPath();
        path.AddCircle(centre.X, centre.Y, radius);
        return path;
    }

    private void DrawGround(SKCanvas canvas, float width, float height, float horizon)
    {
        using var shader = SKShader.CreateLinearGradient(
            new SKPoint(0, horizon),
            new SKPoint(0, height),
            new[] { GroundFar, GroundNear },
            SKShaderTileMode.Clamp);

        _fill.Shader = shader;
        _fill.Color = SKColors.White;
        canvas.DrawRect(SKRect.Create(0, horizon, width, height - horizon), _fill);
        _fill.Shader = null;
    }

    /// <summary>The lines running away to the vanishing point.</summary>
    private void DrawRays(SKCanvas canvas, float width, float height, float horizon)
    {
        _line.Color = Cyan.WithAlpha(191);

        for (var i = -VaporwaveGrid.Rays; i <= VaporwaveGrid.Rays; i++)
        {
            var (atHorizon, atBottom) = VaporwaveGrid.Ray(i);

            canvas.DrawLine(
                (width * 0.5f) + (atHorizon * width), horizon,
                (width * 0.5f) + (atBottom * width), height,
                _line);
        }
    }

    /// <summary>
    /// The rungs rushing toward the viewer, at exactly one per second.
    /// </summary>
    /// <remarks>
    /// Driven off the fractional part of the clock rather than an accumulating
    /// total, so the wrap is exact. Integrate a delta instead and it drifts,
    /// which shows as a hitch once a second for as long as the saver runs.
    /// </remarks>
    private void DrawRungs(SKCanvas canvas, float width, float height, float horizon, double phase)
    {
        var cycle = VaporwaveGrid.Cycle(phase);
        var span = height - horizon;

        for (var k = 0; k < VaporwaveGrid.Rungs; k++)
        {
            var rung = VaporwaveGrid.RungAt(k, cycle);
            if (rung.Alpha <= 0.004f) continue;

            _line.Color = Cyan.WithAlpha((byte)Math.Clamp(rung.Alpha * 255f, 0f, 255f));

            var y = horizon + (rung.DownFromHorizon * span);
            canvas.DrawLine(0, y, width, y, _line);
        }
    }

    private void DrawGlow(SKCanvas canvas, SKRect rect, float side)
    {
        var outer = side * 1.15f;

        using var shader = SKShader.CreateRadialGradient(
            new SKPoint(rect.MidX, rect.MidY), outer,
            new[] { Cyan.WithAlpha(128), Cyan.WithAlpha(0) },
            new[] { side * 0.5f / outer, 1f },
            SKShaderTileMode.Clamp);

        _fill.Shader = shader;
        _fill.Color = SKColors.White;
        canvas.DrawCircle(rect.MidX, rect.MidY, outer, _fill);
        _fill.Shader = null;
    }

    /// <summary>
    /// A bright thin core inside a wider saturated stroke, which is the cheap
    /// way to draw neon in two dimensions.
    /// </summary>
    private void DrawNeonFrame(SKCanvas canvas, SKRect rect)
    {
        _line.StrokeWidth = 3f;
        _line.Color = Magenta;
        canvas.DrawRect(rect, _line);

        _line.StrokeWidth = 1f;
        _line.Color = Cyan;
        canvas.DrawRect(rect, _line);

        _line.StrokeWidth = 1.4f;
    }

    /// <summary>
    /// The title, drawn three times in three colours slightly apart.
    /// </summary>
    /// <remarks>
    /// A fake of the colour fringing you get from a misaligned screen. Cyan
    /// pushed one way, magenta the other, white in the middle. It costs two
    /// extra text draws a frame and it is most of what makes the style read as
    /// the era it is imitating.
    /// </remarks>
    private void DrawText(
        SKCanvas canvas, float width, float height, SKRect rect, int index, ArtPalette palette)
    {
        var (title, artist, _) = FeaturedText.For(_data, index);

        var columnWidth = width * 0.8f;
        var left = (width * 0.5f) - (columnWidth / 2f);
        var top = rect.Bottom + (height * 0.03f);
        var offset = Math.Max(2f, height * 0.004f);

        var shout = title.ToUpperInvariant();
        var size = TextLayout.Fitted(
            shout, columnWidth, height * 0.11f, Math.Max(20f, height * 0.052f), bold: true);

        using var fringe = TextLayout.Paint(size, true, Cyan.WithAlpha(230));
        TextLayout.Draw(canvas, shout, left - offset, top, columnWidth, fringe, centred: true, tracking: 4f);

        fringe.Color = Magenta.WithAlpha(230);
        TextLayout.Draw(canvas, shout, left + offset, top, columnWidth, fringe, centred: true, tracking: 4f);

        fringe.Color = SKColors.White;
        var used = TextLayout.Draw(canvas, shout, left, top, columnWidth, fringe, centred: true, tracking: 4f);

        // The one line the album is allowed to colour.
        using var artistPaint = TextLayout.Paint(
            Math.Max(12f, height * 0.024f), false, palette.Accent.ToSkia());

        TextLayout.Draw(
            canvas, artist.ToUpperInvariant(), left, top + used + (height * 0.014f),
            columnWidth, artistPaint, centred: true, tracking: 6f);
    }

    public void Dispose()
    {
        _fill.Dispose();
        _line.Dispose();
    }
}
